src/datomish/db.cljc

```diff
@@ -12,31 +12,599 @@
+           {:error :transact/bad-entity :entity entity })
+
+    (nil? a)
+    (raise "Bad entity: nil a in " entity
+           {:error :transact/bad-entity :entity entity })
+
+    (nil? v)
+    (raise "Bad entity: nil v in " entity
+           {:error :transact/bad-entity :entity entity })
+
+    (some? rest)
+    (raise "Bad entity: too long " entity
+           {:error :transact/bad-entity :entity entity })
+
+    true
+    entity))
+
+(defn- tx-instant? [db [op e a & _]]
+  (and (= op :db/add)
+       (= e (get-in db [:idents :db/tx]))
+       (= a (get-in db [:idents :db/txInstant]))))
+
+(defn- update-txInstant [db report]
+  "Extract [:db/add :db/tx :db/txInstant ...], and update :txInstant with that value."
+  {:pre [(db? db) (report? report)]}
+
+  ;; TODO: be more efficient here: don't iterate all entities.
+  (if-let [[_ _ _ txInstant] (first (filter (partial tx-instant? db) (:entities report)))]
+    (assoc report :txInstant txInstant)
+    report))
+
+(defn preprocess [db report]
+  {:pre [(db? db) (report? report)]}
+
+  (let [initial-es (or (:entities report) [])
+        ;; :db/tx is a "dynamic enum ident" that maps to the current transaction ID.  This approach
+        ;; mimics DataScript's :db/current-tx.  (We don't follow DataScript because
+        ;; current-txInstant is awkward.)  It's much simpler than Datomic's approach, which appears
+        ;; to unify all id-literals in :db.part/tx to the current transaction value, but also seems
+        ;; inconsistent.
+        tx         (:tx report)
+        db*        (assoc-in db [:idents :db/tx] tx)]
+    (when-not (sequential? initial-es)
+      (raise "Bad transaction data " initial-es ", expected sequential collection"
+             {:error :transact/syntax, :tx-data initial-es}))
+
+    ;; TODO: find an approach that generates less garbage.
+    (->
+      report
+
+      ;; Normalize Datoms into :db/add or :db/retract vectors.
+      (update :entities (partial map maybe-datom->entity))
+
+      ;; Explode map shorthand, such as {:db/id e :attr value :_reverse ref},
+      ;; to a list of vectors, like
+      ;; [[:db/add e :attr value] [:db/add ref :reverse e]].
+      (->> (explode-entities (schema db)))
+
+      (update :entities (partial map ensure-entity-form))
+
+      ;; Replace idents with entids where possible, using db* to capture :db/tx.
+      (update :entities (partial map (partial maybe-ident->entid db*)))
+
+      ;; If an explicit [:db/add :db/tx :db/txInstant] is not given, add one.  Use db* to
+      ;; capture :db/tx.
+      (update :entities (fn [entities]
+                          (if (first (filter (partial tx-instant? db*) entities))
+                            entities
+                            (conj entities (tx-entity db report)))))
+
+      ;; Extract the current txInstant for the report.
+      (->> (update-txInstant db*)))))
+
+(defn- lookup-ref? [x]
+  (and (sequential? x)
+       (= (count x) 2)))
```

Comment:
I've added the first check, but in general these look like `[:attr value]` and `value` is arbitrary.